Razzle plugins + razzle-typescript-plugin

lucasterra opened the pull request:
- [x] gatsby-like plugin system (#437).
- [x] `razzle-plugin-typescript` (tests included)
- [x] Change TypeScript example to use the newly created plugin

---------------------------

It works like this:

Using the plugin with the default options

```js
// razzle.config.js

module.exports = {
  plugins: [
    'typescript'
  ],
};
```

Passing custom options to the plugin:

```js
// razzle.config.js

module.exports = {
  plugins: [
    {
      name: 'typescript',
      options: {
        useBabel: true,
        tsLoader: {
          transpileOnly: true,
          experimentalWatchApi: true,
        },
        forkTsChecker: {
          tsconfig: './tsconfig.json',
          tslint: './tslint.json',
          watch: './src',
          typeCheck: true,
        }
      },
    },
  ],
};
```

lucasterra committed:
Add a plugin system

lucasterra committed:
Add razzle-typescript-plugin

lucasterra committed:
Change TypeScript example to use razzle-typescript-plugin

lucasterra committed:
Small refactor to helpers

lucasterra commented:
Ideally, a plugin should also be able to change the `createJestConfig(...)`, but I will leave it for another PR.

lucasterra committed:
Add myself to contributors

jaredpalmer commented:
This looks epic. On my phone right meow, but will dig into the code tomorrow. I have some ideas about the TypeScript plugin (and different default optimizations) which i’ll tack on. I agree about jest config, we def need a modifyJest() and the TypeScript plugin should have ts-jest setup. 


cc @jorilallo

jaredpalmer commented:
As for the official TS plugin, lets get rid of  tslintloader. We should put tsloader into transpile only and experimentalWatchApi. Then add fork-ts-checker-webpack-plugin

lucasterra committed:
Use fork-ts-checker-webpack-plugin instead of tslint-loader

lucasterra commented:
Removed `tslint-loader`.
Added `experimentalWatchApi`, `transpileOnly` and `fork-ts-checker-webpack-plugin`.

Should we add `thread-loader`/`cache-loader` to `ts-loader` too?

chaffeqa commented:
I have been brooding on this subject the last week, I'd like to offer this alternative idea:

Babel 7 has added [babel/preset-typescript](https://www.npmjs.com/package/@babel/preset-typescript).  Basically it just strips the typescript while compiling down.

With that in mind, I really like the idea of giving full compile control back to babel, and use typescript for linting / DX.

FYI the issues I've been running into include:
* Trying to run 1-off scripts locally (needs something like: `"repl": "node -r esm --require ts-node/register/transpile-only ./bin/repl"`)
* Trying to run 1-off scripts on server (same file, but don't want to add typescript to `dependencies` and have to bundle it in a deploy 👎 )
* Fix jest to support `after.js` dynamic imports (bombs out due to something with typescript --> babel)
* Issues with `.mjs` (may not be related, but the fact that I'm using x2 compilers just complicates so much as far as possible issues introduced)

I'm hoping to fork razzle at some point to attempt to try the idea (need to fork to update the babel razzle preset) and will get back to you about the experiment.  Just thought I'd add it here FYI.

jaredpalmer commented:
I actually found perf degradations when using thread and cache loader with the changes. So my gut says to leave them out.

jaredpalmer commented:
We need a full typescript plugin. We may also do a babel-typescript plugin for Babel 7 TS. As you may or may not know, not all features of TS are supported in Babel TS. So it makes sense to split them I think.

jaredpalmer commented on examples/with-typescript/src/App.tsx:
```diff
@@ -1,4 +1,4 @@
-import * as React from 'react';
```
Put this back?

lucasterra replied:
Why use `import * as React` when we have `esModuleInterop`?

jaredpalmer replied:
You right sorry. I got confused for a sec

jaredpalmer commented on examples/with-typescript/src/client.tsx:
```diff
@@ -1,10 +1,10 @@
-import * as React from 'react';
-import * as ReactDOM from 'react-dom';
+import React from 'react';
```
Fix react imports

jaredpalmer replied:
ignore

jaredpalmer commented on examples/with-typescript/src/server.tsx:
```diff
@@ -1,5 +1,5 @@
-import * as express from 'express';
-import * as React from 'react';
+import express from 'express';
```
turn on `esModuleInterop` flag

jaredpalmer commented on examples/with-typescript/tslint.json:
```diff
@@ -26,7 +26,7 @@
     "no-bitwise": true,
     "no-consecutive-blank-lines": true,
-    "no-console": false,
+    "no-console": true,
```
This makes developing tough. We can keep this relaxed.

jaredpalmer commented on packages/razzle-plugin-typescript/index.js:
```diff
@@ -1,21 +1,23 @@
 const { babelLoaderFinder, eslintLoaderFinder } = require('./helpers');
 
 const defaultOptions = {
   useBabel: true,
```
Use babel should be false by default

lucasterra committed:
Make useBabel false by default

lucasterra committed:
Add esModuleInterop, remove console tslint rules

jaredpalmer commented on packages/razzle-plugin-typescript/.eslintrc:
```diff
@@ -0,0 +1,6 @@
+    "jest": true,
+    "node": true
+  }
+}
```
I think we should actually move eslint out of here. I think we ought to move it of the core as well, making a razzle-plugin-eslint.

lucasterra replied:
Oh, this was just to make linting work on the code and test files. It won't be used by plugin users.

lucasterra replied:
But yeah. I agree with removing eslint from the default webpack config.

jaredpalmer replied:
Ahh sorry I just looked at the filename

jaredpalmer commented on packages/razzle-plugin-typescript/README.md:
```diff
@@ -0,0 +1,56 @@
+# razzle-dev-utils
```
Change title to `Razzle Plugin TypeScript`

lucasterra replied:
Copy/paste really is evil. 😈

jaredpalmer commented on packages/razzle-plugin-typescript/README.md:
````diff
@@ -0,0 +1,56 @@
+## Usage in Razzle Projects
+
+```
+npm install --save razzle-plugin-typescript
````
use yarn

jaredpalmer commented on packages/razzle-plugin-typescript/README.md:
````diff
@@ -0,0 +1,56 @@
+};
+```
+
+Passing custom options to the plugin:
````
Make h2 called Options

jaredpalmer commented on packages/razzle-plugin-typescript/helpers.js:
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
```
Let's move this whole file to `razzle-dev-utils/pluginHelpers` because we'll probably reuse this all over the place

lucasterra replied:
Yeah. Good point. 👍

jaredpalmer commented:
This is looking sweeeet

lucasterra committed:
Improve README

lucasterra committed:
Moved loaderFinder to razzle-dev-utils

lucasterra commented:
All done, I guess.

suciuvlad commented:
Looking forward to the merge 👍

romainquellec commented:
Nice work ! 👍
